docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,125 @@
+# Quality of Service in Kubelet
+
+**Author**: Ananya Kumar (@AnanyaKumar)
+
+**Status**: Draft proposal; prototype in progress.
+
+This document presents the design of quality of service tiers for jobs scheduled on Kubelet, describes use cases, and lays out milestones for its development.
+
+## Motivation
+
+The Kubelet allocates resources to pods in a simple way. Right now there aren’t robust mechanisms in place to ensure that pods run reliably on an overloaded system, and nodes are not very efficiently utilized.
+
+In Kubernetes, some pods have limits specified, and some pods do not. **Pods with limits specified are not very well protected from pods without limits specified**. If a pod without an explicit memory limit goes overboard and uses lots of memory, other pods (with resource specifications) might be killed. This is bad, because users often want a way to launch containers that have resources guarantees, and that stay up reliably.
+
+On the other hand in the current system, **if all pods have specified limits, cluster utilization can be poor**. In the current implementation, Kubelet does not support over-committing pods. Very often, pods don’t use all the resources that they request, which leads to a lot of wasted resources. For example, we might have 4 pods, each reserving 1GB of memory in a system with 4GB memory, but only using 500MB of memory. Theoretically, we could slot in another 4 pods of this type, but Kubernetes will reject new pods.
+
+In this design document, we propose **implementing two tiers of pods** - one in which we guarantee resources (top-tier pods) and the second wherein we provide resources if they’re available and at a slightly lower quality (best-effort pods). Adding a best-effort tier of pods will increase node’s utilization in a Kubernetes cluster (in Borg, supporting best-effort pods increased utilization by  about 20%). We also implement **a policy of automatically classifying pods that don’t provide limits as best-effort**. This provides much better resource guarantees to pods with limits specified.
+
+## QoS Functionality
+
+We will introduce two tiers of pods: top-tier pods and best-effort pods. We will not support multiple levels or priorities.
+
+Best-effort pods will be invisible to top-tier pods:
+- **Existence of best-effort pods should not prevent top tier pods from launching**. When deciding whether to launch a top-tier pod, we only look at resource limits of other top-tier pod. Suppose we have a machine with 4GB of memory. Top-tier pod P1 has a 2GB limit, and 4 best-effort pods have a 500MB limit each. Suppose the user launches a new top-tier pod P2 with a 1GB limit. The Kubelet will evict best-effort pods, in unspecified order, to make room for P2. P2 will then be admitted and allocated the requested resources.

```

Comment:
Actually, they are defined in docs/design/resources.md, which this doc should link to.